pkg/csi/service/common/authmanager.go

```diff
@@ -325,17 +338,80 @@ func getDatastoresWithBlockVolumePrivs(ctx context.Context, vc *vsphere.VirtualC
 	return dsURLToInfoMap, nil
 }
 
-// Creates a map of vsan datastores to file service status (enabled/disabled).
+// Creates a map of vsan datastores to file service enabled status.
+// Since only datastores belonging to clusters with vSAN FS enabled and Host.Config.Storage privileges are returned,
+// file service enabled status will be true for them.
 func getDsToFileServiceEnabledMap(ctx context.Context, vc *vsphere.VirtualCenter) (map[string]bool, error) {
 	log := logger.GetLogger(ctx)
 	log.Debugf("Computing the cluster to file service status (enabled/disabled) map.")
-	datacenters, err := vc.ListDatacenters(ctx)
+
+	//Get clusters with vSAN FS enabled and privileges
+	clustersWithPrivAndFS, err := getFSEnabledClustersWithPriv(ctx, vc)
 	if err != nil {
-		log.Errorf("failed to find datacenters from VC: %+v, Error: %+v", vc.Config.Host, err)
+		log.Errorf("failed to get the file service enabled clusters with privileges. error: %+v", err)
 		return nil, err
 	}
 
 	dsToFileServiceEnabledMap := make(map[string]bool)
+	for _, cluster := range clustersWithPrivAndFS {
+		dsMoList, err := getDatastoreMOsFromCluster(ctx, vc, cluster)
+		if err != nil {
+			log.Errorf("failed to get datastores for cluster %q. error: %+v", cluster.Reference().Value, err)
+			return nil, err
+		}
+		// TODO: Also identify which vSAN datastore is management
+		// and which one is a workload datastore to support file volumes on VMC
+		for _, dsMo := range dsMoList {
+			if dsMo.Summary.Type == VsanDatastoreType {
+				dsToFileServiceEnabledMap[dsMo.Info.GetDatastoreInfo().Url] = true
+			}
+		}
+	}
+	return dsToFileServiceEnabledMap, nil
+}
+
+// Creates a map of cluster id to datastore urls.
+// The key is cluster moid with vSAN FS enabled and Host.Config.Storage privilege.
+// The value is a list of vSAN datastore URLs for each cluster.
+func getFSEnabledClusterToDsURLsMap(ctx context.Context, vc *vsphere.VirtualCenter) (map[string][]string, error) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("Computing the map for vSAN FS enabled clusters to datastore URLS.")
+
+	//Get clusters with vSAN FS enabled and privileges
+	clustersWithPrivAndFS, err := getFSEnabledClustersWithPriv(ctx, vc)
+	if err != nil {
+		log.Errorf("failed to get the file service enabled clusters with privileges. error: %+v", err)
+		return nil, err
+	}
+
+	fsEnabledClusterToDsMap := make(map[string][]string)
+	for _, cluster := range clustersWithPrivAndFS {
+		clusterMoID := cluster.Reference().Value
+		dsMoList, err := getDatastoreMOsFromCluster(ctx, vc, cluster)
+		if err != nil {
+			log.Errorf("failed to get datastores for cluster %q. error: %+v", clusterMoID, err)
+			return nil, err
+		}
+		for _, dsMo := range dsMoList {
+			if dsMo.Summary.Type == VsanDatastoreType {
```

Comment:
Can you confirm if we need a not nil check on `dsMo.Summary` ?

Reply:
No it won't be needed because we're fetching `summary` using property collector in `getDatastoreMOsFromCluster` method that's called to fetch `dsMoList`.
I have tested this for non-vsan datastores also and it doesn't panic.